extensions/amp-viralize-player/0.1/amp-viralize-player.js

```diff
@@ -0,0 +1,378 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {VideoAttributes, VideoEvents} from '../../../src/video-interface';
+import {addParamsToUrl} from '../../../src/url';
+import {installVideoManagerForDoc} from '../../../src/service/video-manager-impl';
+import {isFullscreenElement, removeElement} from '../../../src/dom';
+import {parseJson} from '../../../src/json';
+import {setStyle} from '../../../src/style';
+import {userAssert} from '../../../src/log';
+
+/** @private @const {!string} */
+const TAG = 'amp-viralize-player';
+
+/** @private @const {!string} */
+const BASE_URL = 'https://content.viralize.tv';
+
+/**
+ * @implements {../../../src/video-interface.VideoInterface}
+ */
+class AmpViralizePlayer extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {string} */
+    this.zid_ = null;
+
+    /** @private {object} */
+    this.extraParams_ = null;
+
+    /** @private {Element} */
+    this.container_ = null;
+
+    /** @private {Element} */
+    this.iframe_ = null;
+
+    /** @private {Array} */
+    this.preloadCmdQueue_ = [];
+
+    /** @private {boolean} */
+    this.playerReady_ = false;
+  }
+
+  /**
+   * BaseElement overrides
+   */
+
+  /** @override */
+  buildCallback() {
+    this.zid_ = userAssert(
+      this.element.getAttribute('data-zid'),
+      `The data-zid attribute is required for <${TAG}> %s`,
+      this.element
+    );
+    this.extraParams_ = parseJson(
+      this.element.getAttribute('data-extra') || '{}'
+    );
+    this.extraParams_['activation'] = 'click'; // Autoplay is handled through tag attribute
+    if (this.element.hasAttribute(VideoAttributes.AUTOPLAY)) {
+      // If autoplay, immediately force audio off otherwise AMP framework will assume that player
+      // want to start audio on and will not play it
+      this.extraParams_['sound'] = 'never';
```

Comment:
is this necessary? Amp should be call `mute` before starting the video

Reply:
When the viralize player is created, initially audio tag is always muted. When the content.js controller starts, handleAudio function is invoked and if sound is not never, the tag is unmuted and then as you suggest is muted again.
If the audio is unmuted, VPT unmuted event is emitted and AMP implementation:
`https://github.com/Viralize/amphtml/blob/master/src/service/video-manager-impl.js#L921`
assumes that the user interacted with the tag, that is not true.
When the AMP start checking if the video can start:
`https://github.com/Viralize/amphtml/blob/master/src/service/video-manager-impl.js#L800`
assumes that since an interaction occurred, is not necessary to autoplay, but this is not true.

Setting the sound to never, avoid this unuseful operation `muted --> unmuted --> muted` which can be wrongly interpreted by AMP

Reply:
Ok, I've understand and I agree